File: AdventureScript/Expr.cs
using System.Diagnostics;

namespace AdventureScript
{
    internal abstract class Expr
    {
        public abstract TypeDef Type { get; }

        public abstract int Evaluate(GameState game, int[] frame);

        public virtual void SetValue(GameState game, int[] frame, int value)
        {
            throw new NotImplementedException();
        }

        public virtual bool CanSetValue => false;
        public abstract bool HasSideEffects { get; }
        public abstract bool IsConstant { get; }

        public abstract Precedence Precedence { get; }

        public abstract void WriteExpr(GameState game, CodeWriter writer);

        public static void WriteSubExpr(
            GameState game, 
            Expr expr, 
            Precedence outerPrecedence, 
            CodeWriter writer
            )
        {
            if (expr.Precedence < outerPrecedence)
            {
                writer.Write("(");
                expr.WriteExpr(game, writer);
                writer.Write(")");
            }
            else
            {
                expr.WriteExpr(game, writer);
            }
        }

        protected void WriteSubExpr(GameState game, Expr expr, CodeWriter writer)
        {
            WriteSubExpr(game, expr, this.Precedence, writer);
        }

        public int EvaluateConst(GameState game)
        {
            Debug.Assert(this.IsConstant);
            return Evaluate(game, m_emptyFrame);
        }

        static readonly int[] m_emptyFrame = new int[0];
    }
}

File: rust/adventure_script/src/expr.rs
use std::string::String;
use crate::adventure_script_types::*;
use crate::type_def::*;

pub struct GameState {

}

#[derive(Debug, PartialEq, PartialOrd)]
pub enum Precedence {
    None,
    Ternary,
    AndOr,
    Compare,
    AddSub,
    MulDiv,
    UnaryNegative,
    Member,
    Atomic
}

pub trait Expr {
    fn get_type(&self) -> &TypeRef;
    fn evaluate(&self, game : &GameState, frame : &mut [i32]) -> i32;
    fn evaluate_const(&self, game : &GameState) -> i32;
    fn set_value(&self, game : &GameState, frame : &mut [i32], value : i32);
    fn can_set_value(&self) -> bool;
    fn has_side_effects(&self) -> bool;
    fn is_constant(&self) -> bool;
    fn get_precedence(&self) -> Precedence;
}

pub struct LiteralExpr {
    return_type : TypeRef,
    value : i32
}

impl Expr for LiteralExpr {
    fn get_type(&self) -> &TypeRef {
        &self.return_type
    }
    fn evaluate(&self, _game : &GameState, _frame : &mut [i32]) -> i32 {
        self.value
    }
    fn evaluate_const(&self, _game : &GameState) -> i32 {
        self.value
    }
    fn set_value(&self, _game : &GameState, _frame : &mut [i32], _value : i32) {
        panic!();
    }
    fn can_set_value(&self) -> bool {
        false
    }
    fn has_side_effects(&self) -> bool {
        false
    }
    fn is_constant(&self) -> bool {
        true
    }
    fn get_precedence(&self) -> Precedence {
        Precedence::Atomic
    }
}
